### PR TITLE
Add support for ucmp/scmp

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch adds support for llvm.ucmp/scmp intrinsics.
See https://llvm.org/docs/LangRef.html#llvm-scmp-intrinsic

### CODE REVIEW

**Comment by dtcxzyw** [comment]
cc @nikic @Poseydon42

**Review by nikic — COMMENTED** [review]
Looks right to me.

File: tools/alive_parser.cpp [hunk 1]
```diff
@@ -754,6 +759,12 @@ static unique_ptr<Instr> parse_binop(string_view name, token op_token) {
   case SMIN: op = BinOp::SMin; break;
   case SMAX: op = BinOp::SMax; break;
   case ABS:  op = BinOp::Abs; break;
+  case UCMP:
```
Comment (nunoplopes) [inline tools/alive_parser.cpp:762]: please following the exist style

File: ir/instr.cpp [hunk 2]
```diff
@@ -184,6 +184,12 @@ void BinOp::print(ostream &os) const {
   case SMin:          str = "smin "; break;
   case SMax:          str = "smax "; break;
   case Abs:           str = "abs "; break;
+  case UCmp:
```
Comment (nunoplopes) [inline ir/instr.cpp:187]: please follow the style

File: llvm_util/llvm2alive.cpp [hunk 3]
```diff
@@ -921,6 +922,12 @@ class llvm2alive_ : public llvm::InstVisitor<llvm2alive_, unique_ptr<Instr>> {
       case llvm::Intrinsic::smin:     op = BinOp::SMin; break;
       case llvm::Intrinsic::smax:     op = BinOp::SMax; break;
       case llvm::Intrinsic::abs:      op = BinOp::Abs; break;
+      case llvm::Intrinsic::ucmp:
```
Comment (nunoplopes) [inline llvm_util/llvm2alive.cpp:925]: please follow the style

**Review by nunoplopes — CHANGES_REQUESTED** [review]
functionality looks good. Just the style doesn't match in a couple of places.

File: ir/instr.cpp [hunk 4]
```diff
@@ -453,6 +458,30 @@ StateValue BinOp::toSMT(State &s) const {
...
+        break;
+      case SCmp:
+        lt = a.slt(b);
+        gt = a.sgt(b);
+        break;
+      default:
+        UNREACHABLE();
+      }
+      uint32_t resBits = getType().bits();
+      return {expr::mkIf(std::move(lt), expr::mkInt(-1, resBits),
```
Comment (nunoplopes) [inline ir/instr.cpp:479]: one suggestion (sorry, just occurred to me now).
it's more efficient if done like this:
mkIf(a == b, 0, mkIf(lt, -1, 1))

No need for gt and == is way more efficient.

File: ir/instr.cpp [hunk 5]
```diff
@@ -453,6 +458,18 @@ StateValue BinOp::toSMT(State &s) const {
       return { a.abs(), ap && bp && (b == 0 || a != expr::IntSMin(a.bits())) };
     };
     break;
+
+  case UCmp:
+  case SCmp:
+    fn = [&](auto &a, auto &ap, auto &b, auto &bp) -> StateValue {
+      uint32_t resBits = getType().bits();
```
Comment (dtcxzyw) [inline ir/instr.cpp:465]: Sorry, I find that it doesn't work with vector types:
```
; ./alive2-tv test.ll --disable-undef-input
define <2 x i8> @src(<2 x i32> range(i32 0, 10) %x, <2 x i32> range(i32 0, 10) %y) {
  %cmp = call <2 x i8> @llvm.scmp.v2i32.v2i8(<2 x i32> %x, <2 x i32> %y)
  ret <2 x i8> %cmp
}

define <2 x i8> @tgt(<2 x i32> %x, <2 x i32> %y) {
  %cmp = call <2 x i8> @llvm.ucmp.v2i32.v2i8(<2 x i32> %x, <2 x i32> %y)
  ret <2 x i8> %cmp
}

----------------------------------------
define <2 x i8> @src(<2 x i32> %x, <2 x i32> %y) {
init:
  %#range_%x = !range <2 x i32> %x, i32 0, i32 10
  %#range_%y = !range <2 x i32> %y, i32 0, i32 10
  br label %#0

#0:
  %cmp = scmp <2 x i8> <2 x i32> %#range_%x, %#range_%y
  ret <2 x i8> %cmp
}
=>
define <2 x i8> @tgt(<2 x i32> %x, <2 x i32> %y) {
#0:
  %cmp = ucmp <2 x i8> <2 x i32> %x, %y
  ret <2 x i8> %cmp
}
Transformation doesn't verify!

ERROR: Value mismatch

Example:
<2 x i32> %x = < #x00000000 (0), #x0000008a (138) >
<2 x i32> %y = < #x00000000 (0), #x00000000 (0) >

Source:
<2 x i32> %#range_%x = < #x00000000 (0), poison >
<2 x i32> %#range_%y = < #x00000000 (0), #x00000000 (0) >
UB triggered on br

Target:
<2 x i8> %cmp = < #x00 (0), #x01 (1) >
Source value: < #xff (255, -1), poison >
Target value: < #x00 (0), #x01 (1) >
```

Comment (dtcxzyw) [reply ir/instr.cpp:465]: I will post a fix.

**Pull request merged**